packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,500 @@
+        :param host: RESTful API hostname / IP address
+        :param port: RESTful API port number
+        :param api_spec_path: Directory API path and filename of the API spec YAML source file.
+        :param timeout_window: the timeout (in seconds) for a request to be handled.
+        """
+        self.address = address
+        self.host = host
+        self.port = port
+        self.connection_id = connection_id
+        self.restricted_to_protocols = restricted_to_protocols
+        self.in_queue = None  # type: Optional[asyncio.Queue]
+        self.loop = None  # type: Optional[asyncio.AbstractEventLoop]
+        self.thread = None  # type: Optional[Thread]
+        self.lock = Lock()
+        self.is_stopped = True
+        self._api_spec = APISpec(api_spec_path)
+        self.timeout_window = timeout_window
+        self.http_server = None  # type: Optional[HTTPServer]
+        self.dispatch_ready_envelopes = {}  # type: Dict[RequestId, Envelope]
+        self.timed_out_request_ids = set()  # type: Set[RequestId]
+
+    @property
+    def api_spec(self) -> APISpec:
+        """Get the api spec."""
+        return self._api_spec
+
+    def connect(self):
+        """
+        Connect.
+
+        Upon HTTP Channel connection, kickstart the HTTP Server in its own thread.
+        """
+        if self.is_stopped:
+            try:
+                self.http_server = HTTPServer(
+                    (self.host, self.port), HTTPHandlerFactory(self)
+                )
+                logger.info("HTTP Server has connected to port: {}.".format(self.port))
+                self.thread = Thread(target=self.http_server.serve_forever)
+                self.thread.start()
+                self.is_stopped = False
+            except OSError:
+                logger.error(
+                    "{}:{} is already in use, please try another Socket.".format(
+                        self.host, self.port
+                    )
+                )
+
+    def send(self, envelope: Envelope) -> None:
+        """
+        Send the envelope in_queue.
+
+        :param envelope: the envelope
+        :return: None
+        """
+        assert self.http_server is not None, "Server not connected, call connect first!"
+
+        if envelope.protocol_id not in self.restricted_to_protocols:
+            logger.error(
+                "This envelope cannot be sent with the http connection: protocol_id={}".format(
+                    envelope.protocol_id
+                )
+            )
+            raise ValueError("Cannot send message.")
+
+        if envelope.to in self.timed_out_request_ids:
+            self.timed_out_request_ids.remove(envelope.to)
+            logger.warning(
+                "Dropping envelope for request id {} which has timed out.".format(
+                    envelope.to
+                )
+            )
+        else:
+            self.dispatch_ready_envelopes.update({envelope.to: envelope})
```

Comment:
This should be executed only when the agent handler produces a response to a previous request. However, it seems we can dispatch a message even if it is not associated to a request id (although it will never be consumed by the `get_response` method).

So I propose to also check that there is a pending request to be answered.

Reply:
Done now via `self.pending_request_ids`.